### ISSUE TITLE
Maybe S3249 should ignore `return base.GetHashCode()`

### ISSUE THREAD

**Issue by sm-g** [SonarSource/sonar-csharp#1792]
Or how should I rewrite this code to satisfy S3249?

```cs
        public override int GetHashCode()
        {
            if (IsTransient())
            {
                return base.GetHashCode();
            }
            return Id.GetHashCode();
        }
```

sample taken from 
https://github.com/dotnet-architecture/eShopOnContainers/blob/dev/src/Services/Ordering/Ordering.Domain/SeedWork/Entity.cs#L76

**Comment by Evangelink** [comment]
Hi @sm-g,

Thanks for the feedback. I think the rule is working properly, it's never a good idea to return `base.GetHashCode()` when the class directly inherits from `object` because then the equality of 2 objects is only allowed when there is a reference equality.

Also you might have noticed a different warning on `Id.GetHashCode()` because it's recommended not to rely on mutable elements to calculate the hash. Otherwise if your object is contained in a data structure (let's say key to a `Dictionary`) you can mutate the object by changing the `Id` which will invalidate the equality between objects.

I recommend reading some blogs about `GetHashCode` because it's more tricky than what we usually think of. For example you can start by this article from Eric Lippert: https://blogs.msdn.microsoft.com/ericlippert/2011/02/28/guidelines-and-rules-for-gethashcode/

**Comment by sm-g** [comment]
So if I need exactly such behavior (compare by reference) I must disable warning?

**Comment by Evangelink** [comment]
You can mark the instance as `Won't Fix` on your SonarQube/SonarCloud project.

**Comment by sm-g** [comment]
@Evangelink another false positive for this, when I want apply some attributes to method but still call base implementation:

```
        [EditorBrowsable(EditorBrowsableState.Never)]
        public override bool Equals(object obj) => base.Equals(obj);
```
